### PR TITLE
Card #2771 - Add keyboard navigation

### PR DESCRIPTION
**Pull request by ediblecode**
https://nice-digital.mingle.thoughtworks.com/projects/problem_management/cards/2771

Add ability to use arrow keys to navigate within menus, home and end keys,
space/enter to activate and escape to close menus. Add focus ring style.

### CODE REVIEW

File: lib/utils/dom.js [hunk 1]
```diff
@@ -2,6 +2,184 @@ var config = require('../config');
...
+
+	return labelledBy;
+};
+
+/**
+* If the element passed in is accessible then true is retruned.
+ * @param      {HTMLElement} element
+ * @returns    {boolean} { boolean }
+ */
+function isNavigable(element){
```
Comment (ediblecode) [inline lib/utils/dom.js]: Should these be `utils.` or are they meant to be private functions to this module?

File: lib/utils/dom.js [hunk 2]
```diff
@@ -2,6 +2,184 @@ var config = require('../config');
...
+			var nextControllingEl = utils.closestWithClass(controllingEl, "menu");
+			if(nextControllingEl)
+				return utils.getControllingElementId(nextControllingEl);
+	}
+
+	return labelledBy;
+};
+
+/**
+* If the element passed in is accessible then true is retruned.
```
Comment (ediblecode) [inline lib/utils/dom.js]: "is accessible" should probably be "tabble" or "focussable" or something I think!

File: lib/utils/dom.js [hunk 3]
```diff
@@ -2,6 +2,184 @@ var config = require('../config');
...
+};
+
+/**
+ * Getst the element controlling the element passed in. If the controlling element is
+ *not accessible find the controlling element.
+ * @param      {HTMLElement} element you are trying to find the .
+ * @returns    {HTMLElementId} { Id }
+ */
+
+//passing in main-menu, splits out the same thing... why?
```
Comment (ediblecode) [inline lib/utils/dom.js]: Please remove/rewrite this comment

File: lib/utils/dom.js [hunk 4]
```diff
@@ -2,6 +2,184 @@ var config = require('../config');
...
+	var elControls = document.getElementById(elId).getAttribute("aria-controls");
+	if(!elControls)
+		return null;
+	return elControls.split(" ");
+};
+
+/**
+ * Getst the element controlling the element passed in. If the controlling element is
+ *not accessible find the controlling element.
+ * @param      {HTMLElement} element you are trying to find the .
```
Comment (ediblecode) [inline lib/utils/dom.js]: Please re-write the description for this parameter

File: lib/utils/dom.js [hunk 5]
```diff
@@ -2,6 +2,184 @@ var config = require('../config');
...
+
+/**
+*gets the index of the next accessible element within a list of elements in relation to direction
+ * @param      {int}int  current index of element
+ * @param      {Array} list of elements
+ * @param      {int} direction which the index is trying to move -1 being the previous index
+ * 1 being the nect index
+ * @returns    {int} { index }
+ */
+function skipDisabledLinks(index,links, skipDirectionInt){
```
Comment (ediblecode) [inline lib/utils/dom.js]: Should be on the utils object

File: lib/utils/dom.js [hunk 6]
```diff
@@ -2,6 +2,184 @@ var config = require('../config');
...
+		index += skipDirectionInt;
+		index = wrapToEnd(index, links);
+	}
+	return index;
+}
+
+/**
+ * Gets the links from within the container, as an Array.
+ *
+ * @param      {HTMLElement}  container    The containing element
```
Comment (ediblecode) [inline lib/utils/dom.js]: This comment needs updating to match the methos - seems to be a list of containers not a single container

File: lib/utils/dom.js [hunk 7]
```diff
@@ -2,6 +2,184 @@ var config = require('../config');
...
+ * @return     {Array}  The links within the container.
+ */
+utils.getLinks = function(containers) {
```
Comment (ediblecode) [inline lib/utils/dom.js]: `containers` should maybe be `containerIds`

File: lib/events/index.js [hunk 8]
```diff
@@ -107,22 +176,35 @@ function clickhandler( ev ) {
...
 }
 
+function focusOnMenu(controllerId){
+	var controls = domUtils.getIdsOfControlledContainers(controllerId);
+	var firstControlledContainer = document.getElementById(controls[0]);
+	var menu = firstControlledContainer.querySelectorAll('.menu')[0] || firstControlledContainer;
+
+	menu.focus();
+
+	return;
```
Comment (ediblecode) [inline lib/events/index.js]: This doesn't need a `return;` here

File: lib/events/index.js [hunk 9]
```diff
@@ -107,22 +176,35 @@ function clickhandler( ev ) {
...
+    var elementId = cleanClass( target.className );
+    switch (elementId) {
         case 'menu-evidence':
             this.state.toggleEvidence();
+            focusOnMenu(elementId);
             break;
 
         case 'menu-profile':
             this.state.toggleProfile();
+            focusOnMenu(elementId) ;
```
Comment (ediblecode) [inline lib/events/index.js]: Please remove the space before the semicolon

File: lib/events/index.js [hunk 10]
```diff
@@ -13,23 +30,75 @@ function tophatEvents( document, tophatElement, serviceElement, config ) {
...
+			domUtils.getLinks(menuIdsWithinScope)[0].focus();
+			break;
+		case KeysDict.end:
+			var links = domUtils.getLinks(menuIdsWithinScope);
+			links[links.length - 1].focus();
+			break;
+		default:
+			return;
+	}
+	e.preventDefault();
```
Comment (ediblecode) [inline lib/events/index.js]: Does e.preventDefault work in ie8? There's a `utils.enhanceEvent` function that seems to do some cross browser prevent default stuff which I wonder if we should be using.

File: lib/utils/dom.js (hunk 1 above)
Comment (1996dylanriley) [reply lib/utils/dom.js]: I made them private as they are only intended to be used within utils.

**Pull request merged**